File: haskell/Problem33.hs
-- Copyright 2015 Mitchell Kember. Subject to the MIT License.
-- Project Euler: Problem 33
-- Digit cancelling fractions

module Problem33 where

import Common (digits, undigits)

import Data.List (intersect, (\\))
import Data.Maybe (mapMaybe)
import Data.Ratio (Ratio, denominator, (%))

curious :: (Int, Int) -> Maybe (Ratio Int)
curious (a, b)
    | not trivial && not invalid && r == r' = Just r'
    | otherwise = Nothing
  where
    da = digits a
    db = digits b
    common = intersect da db
    trivial = null common || common == [0]
    a' = undigits $ da \\ common
    b' = undigits $ db \\ common
    invalid = a' == 0 || b' == 0
    r = a % b
    r' = a' % b'

solve :: Int
solve = denominator . product . mapMaybe curious $ pairs
  where
    pairs = [(a, b) | b <- [10..99], a <- [10..b-1]]

File: c++/src/problem_33.cpp
// Copyright 2016 Mitchell Kember. Subject to the MIT License.
// Project Euler: Problem 33
// Digit cancelling fractions

#include "common.hpp"

namespace problem_33 {

bool is_curious(const long num, const long den) {
	const long num_d1 = num % 10;
	const long num_d2 = num / 10;
	const long den_d1 = den % 10;
	const long den_d2 = den / 10;

	if (num_d1 == 0 || den_d1 == 0) {
		return false;
	}
	if (num_d1 == den_d1) {
		return num * den_d2 == den * num_d2;
	}
	if (num_d1 == den_d2) {
		return num * den_d1 == den * num_d2;
	}
	if (num_d2 == den_d1) {
		return num * den_d2 == den * num_d1;
	}
	if (num_d2 == den_d2) {
		return num * den_d1 == den * num_d1;
	}
	return false;
}

long solve() {
	long num = 1;
	long den = 1;
	for (long d = 12; d <= 99; ++d) {
		for (long n = 11; n < d; ++n) {
			if (is_curious(n, d)) {
				num *= n;
				den *= d;
			}
		}
	}
	common::reduce_fraction(num, den);
	return den;
}

} // namespace problem_33
